tests/test_ga4_conversion_window.py

```diff
@@ -0,0 +1,168 @@
+                {"id": self.custom_report_id_2, "name": "Test Report 2"},
+            ]
+        }
+        if original:
+            return return_value
+
+        if self.start_date:
+            return_value["start_date"] = self.start_date
+        if self.request_window_size:
+            return_value["request_window_size"] = self.request_window_size
+        return return_value
+
+
+    def streams_to_test(self):
+        # testing all streams creates massive quota issues
+        custom_id = self.custom_reports_names_to_ids()['Test Report 1']
+        return {
+            'content_group_report',
+            custom_id
+        }
+
+
+    def run_test(self):
+        """
+        Testing that basic sync functions without Critical Errors when
+        a valid conversion_window is set.
+        """
+
+        # Create a connection
+        conn_id = connections.ensure_connection(self, original_properties=False)
+
+        # Run a discovery job
+        found_catalogs = self.run_and_verify_check_mode(conn_id)
+
+        # Select tables and fields
+        test_catalogs = [catalog
+                         for catalog in found_catalogs
+                         if catalog['stream_name'] in self.streams_to_test()]
+        streams_to_selected_fields = {stream: set() for stream in self.streams_to_test()}
+        self.select_streams_and_fields(conn_id, test_catalogs, streams_to_selected_fields)
+
+        # set state to ensure conversion window is used
+        today_datetime = dt.strftime(dt.utcnow(), self.BOOKMARK_FORMAT)
+        initial_state = {
+            'bookmarks': {
+                stream: { os.getenv('TAP_GA4_PROPERTY_ID'): {'last_report_date': today_datetime}}
+                for stream in self.streams_to_test()
+            }, 'currently_syncing': None
+        }
+
+        menagerie.set_state(conn_id, initial_state)
+
+        # Run a sync
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        # Verify the tap and target do not throw a critical error
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # Verify tap replicates through today by check state
+        final_state = menagerie.get_state(conn_id)
+        self.assertDictEqual(final_state, initial_state)
+
+        # Verify converstion window is used
+        synced_messages_by_stream = runner.get_records_from_target_output()
+        for stream in self.streams_to_test():
+            with self.subTest(stream=stream):
+
+                stream_name = self.get_stream_name(stream)
+                expected_replication_key = next(iter(self.expected_metadata().get(stream_name).get(self.REPLICATION_KEYS)))
+                if not expected_replication_key:
+                    LOGGER.warn(f"********** Failed to get replication key for stream: {stream_name}")
+                    continue
```

Comment:
Should this fail the test.  Is there ever a time we don't have a replication key?  If this is for full table replication maybe we should take those tests out of streams to test if they aren't applicable.  But this looks like we could be hiding a failure or setup issue and then not test a stream we intend to.

Reply:
I agree, this should be a hard fail to fix the expectations in base or otherwise resolve the conflict.  Will change this to an assert.